### PR TITLE
Pipes - Sairagul - MediaRanker

### PR DESCRIPTION
**Pull request by sairagula**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | I added a few custom methods to my Work model. work_with_the_most_votes method is one of them. It will return a work with the most votes
Describe how you approached testing that model method. What edge cases did you come up with? |Testing these custom methods were complicated for me, I used data from yml file as expected value and tested if result was equal to expected 
Describe an edge case test you wrote for a controller | I tested login method with an edge case where it will through an error if no name given
What are `session` and `flash`? What is the difference between them? | These are special objects which behave like hashes. They are used to save data across multiple requests. A flash stores a value until the next request, while a session stores data during the complete session. 
Describe a controller filter you wrote. | To DRY up work controller's show, edit, update, destroy methods , I created find_work_by_params_id  private method, that finds a work from the params id key or throws an error if not found .
What was one thing that you gained more clarity on through this assignment? | A lot of things are still kind of foggy in my head, but now I know the whole structure of the application, and what  parts it needs to function. I enjoyed using filters, sessions and flash. 
What is the Heroku URL of your deployed application | https://sairamediaranker.herokuapp.com/
Do you have any recommendations on how we could improve this project for the next cohort? |  I still didn't digest most of the new things we learned in class last week, I didn't have time to go through them , understand and use in my project. So probably longer time period for the project would be a good practice.

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  # get "/works/:id", to: "works#show", as: "work"
+  #
+  # get "/work/:id/edit", to: "works#edit", as: "edit_work"
+  # patch "/works/:id", to: "works#update"
+  #
+  # delete "/work/:id", to: "works#destroy"
+  root :to => "main#index"
+  post "works/:id/upvote", to: "works#upvote", as:"upvote"
+  resources :works
+  resources :users
```
Comment (droberts-sea) [inline config/routes.rb:19]: You probably don't need the full `resources :users` here. Make sure to only generate routes you're planning to use.

File: app/models/work.rb [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  validates :title, presence: true, uniqueness: { message: "The title has already been taken" }
+  validates :publication_year, numericality: { only_integer: true }, allow_nil: true
+
+  def self.work_with_the_most_votes
+    return Work.left_outer_joins(:votes).select('works.*, COUNT(votes.*) AS votes_count').group('works.id').order('votes_count desc').first
+  end
+
+  def self.top_n_in_category(category, count)
+    return Work.left_outer_joins(:votes).select('works.*, COUNT(votes.*) AS votes_count').group('works.id').where(category: category).order('votes_count desc').limit(count)
+  end
```
Comment (droberts-sea) [inline app/models/work.rb:13]: Good work figuring out these database operations.

File: test/models/work_test.rb [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+      res.must_include Work.find_by_title("Countdown to Ecstasy 2")
+      res.must_include Work.find_by_title("Countdown to Ecstasy 4")
+      res.size.must_equal 3
+    end
+
+    it "can find top 3 in movies" do
+      res = Work.top_n_in_category(:movie, 10)
+      res.must_include Work.find_by_title("Jurassic_Park")
+      res.size.must_equal 1
+    end
```
Comment (droberts-sea) [inline test/models/work_test.rb:54]: I would also be interested to see:
- What happens if there are no works?
- What if there are more than n works?

File: test/models/work_test.rb [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+    it "returns all works ordered by vote count" do
+      res = Work.all_with_vote_count
+      expected = [
+        Work.find_by_title("Countdown to Ecstasy"),
```
Comment (droberts-sea) [inline test/models/work_test.rb:59]: This test ends up being very fragile - if your fixtures change, this test will probably break.

Instead of requiring specific works to be returned, you might check that the right number of works are there, and that each work has a vote count equal to or less than the vote before it.

File: test/controllers/works_controller_test.rb [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+
+    it "should return not found when given invalid work ID" do
+      bad_w_id = Work.last.id + 1 # Arrange
+      work_count = Work.count
+
+      delete work_path(bad_w_id) #Act
+
+      must_respond_with :not_found # Assert
+      Work.count.must_equal work_count
+    end
```
Comment (droberts-sea) [inline test/controllers/works_controller_test.rb:153]: What if you try to destroy a work with associated votes? Does it work? What happens to the votes?

**Comment by droberts-sea** [comment]
# Media Ranker
## What We're Looking For

Feature | Feedback
---     | ---
**Core Requirements** |
Git hygiene | yes
Comprehension questions	| yes
**General** |
Rails fundamentals (RESTful routing, use of named paths) | yes - some extra routes
Semantic HTML | yes
Errors are reported to the user | yes
Business logic lives in the models | yes - good work
Models are thoroughly tested, including relations, validations and any custom logic | yes - see inline comments for suggestions
Controllers are thoroughly tested | yes - good work
**Wave 1 - Media** |
Splash page shows the three media categories | yes
Basic CRUD operations on media are present and functional | yes
**Wave 2 - Users and Votes** |
Users can log in and log out | yes
The ID of the current user is stored in the session | yes
Individual user pages and the user list are present | yes
A user cannot vote for the same media more than once | yes
All media lists are ordered by vote count | yes
Splash page contains a media spotlight | yes
Media pages contain lists of voting users | yes
**Wave 3 - Styling** |
Foundation is used appropriately | yes
Look and feel is similar to the original | yes - good work
**Overall** | Great job overall! This is a large project with many complex requirements, and you've done a great job with it. I'm especially happy with your controller testing. Keep up the hard work!
